stdlib/public/core/String.swift

```diff
@@ -910,41 +883,14 @@ extension String {
       }
     }
 
-    // TODO(String performance): Try out incremental case-conversion rather than
-    // make UTF-16 array beforehand
-    let codeUnits = Array(self.utf16).withUnsafeBufferPointer {
-      (uChars: UnsafeBufferPointer<UInt16>) -> Array<UInt16> in
-      var length: Int = 0
-      let result = Array<UInt16>(unsafeUninitializedCapacity: uChars.count) {
-        buffer, initializedCount in
-        var err = __swift_stdlib_U_ZERO_ERROR
-        length = Int(truncatingIfNeeded:
-          __swift_stdlib_u_strToUpper(
-            buffer.baseAddress._unsafelyUnwrappedUnchecked,
-            Int32(buffer.count),
-            uChars.baseAddress._unsafelyUnwrappedUnchecked,
-            Int32(uChars.count),
-            "",
-            &err))
-        initializedCount = min(length, uChars.count)
-      }
-      if length > uChars.count {
-        var err = __swift_stdlib_U_ZERO_ERROR
-        return Array<UInt16>(unsafeUninitializedCapacity: length) {
-          buffer, initializedCount in
-          __swift_stdlib_u_strToUpper(
-            buffer.baseAddress._unsafelyUnwrappedUnchecked,
-            Int32(buffer.count),
-            uChars.baseAddress._unsafelyUnwrappedUnchecked,
-            Int32(uChars.count),
-            "",
-            &err)
-          initializedCount = length
-        }
-      }
-      return result
+    var result = ""
+    result.reserveCapacity(utf8.count)
+
+    for scalar in unicodeScalars {
+      result += scalar.properties.uppercaseMapping
     }
-    return codeUnits.withUnsafeBufferPointer { String._uncheckedFromUTF16($0) }
+
+    return result
```

Comment:
I wonder if we have good benchmark coverage for this. This could be a nice improvement.